### PR TITLE
Add SQLite WITHOUT ROWID table option

### PR DESCRIPTION
**Pull request by lazerg**
|      Q       |   A
|------------- | -----------
| Type         | feature
| Fixed issues | #7473

#### Summary

SQLite supports the `WITHOUT ROWID` and `STRICT` table modifiers, but there was no way to emit them. This adds two boolean table options, `without_rowid` and `strict`, that append the corresponding modifiers to the generated `CREATE TABLE` statement, the same way PostgreSQL's `unlogged` option does.

### CODE REVIEW

**Comment by ausi** [comment]
Should we add fetching of these options in `SQLiteSchemaManager::fetchTableOptionsByTable()` as well?

Since SQLite 3.37.0 this can be retrieved using [`PRAGMA table_list`](https://sqlite.org/pragma.html#pragma_table_list). For older versions we probably need to fallback to parsing the `CREATE TABLE`.

**Comment by derrabus** [comment]
> Should we add fetching of these options in `SQLiteSchemaManager::fetchTableOptionsByTable()` as well?

Yes, as long we don't detect those options when intropecting a schema, this feature in incomplete. We would basically always erase those options when diffing a schema. Also, we need functional tests that show us that the generated DDL statements actually work.

**Comment by lazerg** [comment]
Introspection now reads `without_rowid` and `strict` back, using `PRAGMA table_list` on SQLite 3.37+ and falling back to parsing the `CREATE TABLE` statement on older versions, so they're no longer erased when diffing. I also added functional tests that create the tables from the generated DDL and introspect the options back to confirm the round-trip.

**Review by ausi — APPROVED** [review]

File: src/Platforms/SQLite/SQLiteMetadataProvider.php [hunk 1]
```diff
@@ -725,13 +726,36 @@ public function getTableOptionsForTable(
      */
     private function getTableOptions(iterable $tableNames): iterable
     {
+        $supportsTableList = version_compare($this->connection->getServerVersion(), '3.37.0', '>=');
```
Comment (derrabus) [inline src/Platforms/SQLite/SQLiteMetadataProvider.php]: We don't check the server version at runtime. Please create a new SQLitePlatform for this version or find a way that works for all supported SQLite versions.

Comment (derrabus) [reply src/Platforms/SQLite/SQLiteMetadataProvider.php]: That being said, the lowest SQLite version that we test against is 3.37.2. I'm not sure if we need to care about anything lower than that.

Comment (lazerg) [reply src/Platforms/SQLite/SQLiteMetadataProvider.php]: Dropped the version check. Both introspection paths now read the modifiers out of the CREATE TABLE statement in sqlite_master, which they already fetch for the table comment. No pragma_table_list query and no version branch is left, so it works on every SQLite version and it saves one query per table.

I went this way instead of a version-specific platform because docs/en/reference/platforms.rst lists `SQLitePlatform` "for all versions", and `AbstractSQLiteDriver::getDatabasePlatform()` ignores the version provider entirely. Parsing sqlite_master is also the existing idiom in both files: comments, unique constraints, collations and AUTOINCREMENT are all read that way.

The functional round-trip tests now exercise the parser against a real SQLite server, and the unit test no longer stubs a version.

**Comment by derrabus** [comment]
Thank you for your well-crafted PR. I've done some research on the feature that we're about to add and would like to split the PR and the main reason for that is the `STRICT` option.

Supporting `WITHOUT ROWID` is absolutely the right call. We could even think about making this the default for the ORM where we require each table to have a proper primary key. The tests are clean and prove that we can deploy and introspect a table with that option. We can merge this feature right away and I wouldn't want discussions on the `STRICT` feature to block it.

Now to the `STRICT` flag. That flag is mostly incompatible with how DBAL operates today because it limits us to the column types that SQLite supports natively, the so-called storage classes. For instance, you won't be able to create a strict table with a `DATETIME` or a `VARCHAR` column. You can easily verify what I'm saying by adding such a column to the table you create in your new `testIntrospectStrictTableOption()` test. On those tables, we'd need to map types in an entirely different way. We've had a discussion on this a couple of years ago that I completely forgot about, see #5149.

Long story short: If we really want to support the strict flag, that feature would be a lot more complex than passing along the flag. I'm not sure if it's worth the effort, tbh. But if you have a use-case for that and are willing to implement full support (as in: all DBAL types have to work with strict mode), we should do that in a separate PR.

**Comment by ausi** [comment]
> Supporting `WITHOUT ROWID` is absolutely the right call. We could even think about making this the default for the ORM where we require each table to have a proper primary key.

I’m not sure about changing the default. For the regular ORM case (an entity with an autoincrement integer primary key) there is no real difference between with and `WITHOUT ROWID`. And for all the other cases it strongly depends on the use case if `WITHOUT ROWID` makes it perform faster or slower and uses more or less storage space. But on the other hand, for better compatibility with other databases, using `WITHOUT ROWID` by default would make sense as most databases work in a similar way AFAIK. If you decide on changing the default, I think it could make sense to wait for next major version to reduce surprises.

> …and would like to split the PR and the main reason for that is the `STRICT` option.

👍

**Comment by lazerg** [comment]
Makes sense, thanks for digging up #5149. I've removed STRICT support from this PR, it's now scoped to WITHOUT ROWID only, ready to merge as-is. Happy to pick up STRICT separately later if there's a concrete use-case worth the type-mapping work.

File: src/Platforms/SQLite/SQLiteMetadataProvider.php [hunk 2]
```diff
@@ -777,6 +782,15 @@ private function parseTableCommentFromSQL(string $table, string $sql): ?string
         return $comment === '' ? null : $comment;
     }
 
+    private function parseTableOptionsFromSQL(string $sql): bool
```
Comment (derrabus) [inline src/Platforms/SQLite/SQLiteMetadataProvider.php]: The name of this function does not fit anymore because it only parses and reports one very specific option.

Comment (lazerg) [reply src/Platforms/SQLite/SQLiteMetadataProvider.php]: Renamed to `parseWithoutRowidFromSQL()`

**Review by ausi — APPROVED** [review]

File: src/Platforms/SQLite/SQLiteMetadataProvider.php [hunk 3]
```diff
@@ -726,12 +726,17 @@ public function getTableOptionsForTable(
...
-            ]);
+            $createSQL = $this->getCreateTableSQL($tableName);
+
+            $options = [
+                'comment' => $this->parseTableCommentFromSQL($tableName, $createSQL),
+            ];
+
+            if ($this->parseWithoutRowidFromSQL($createSQL)) {
+                $options['without_rowid'] = true;
+            }
```
Comment (morozov) [inline src/Platforms/SQLite/SQLiteMetadataProvider.php]: Should it be just an assignment of the return value? Otherwise, only the `true` value will be explicitly expressed. Besides that, the if increases the cyclomatic code complexity for no reason.

File: tests/Functional/Schema/SQLiteSchemaManagerTest.php [hunk 4]
```diff
@@ -285,6 +285,31 @@ public function testOnlyOwnCommentIsParsed(): void
...
+                    ->setUnquotedName('id')
+                    ->setTypeName(Types::INTEGER)
+                    ->create(),
+            )
+            ->setPrimaryKeyConstraint(
+                PrimaryKeyConstraint::editor()
+                    ->setUnquotedColumnNames('id')
+                    ->create(),
+            )
+            ->setOptions(['without_rowid' => true])
```
Comment (morozov) [inline tests/Functional/Schema/SQLiteSchemaManagerTest.php]: It would be good to cover both the `true` and `false` values via `TestWith` in the same test.

File: tests/Platforms/SQLite/SQLiteMetadataProviderTest.php [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+        self::assertCount(1, $rows);
+
+        $options = $rows[0]->getOptions();
+        self::assertSame($expectedWithoutRowid, $options['without_rowid'] ?? false);
+    }
+
+    /** @return iterable<string, array{string, bool}> */
+    public static function tableOptionsProvider(): iterable
+    {
+        yield 'plain' => ['CREATE TABLE mytable (id INTEGER)', false];
```
Comment (morozov) [inline tests/Platforms/SQLite/SQLiteMetadataProviderTest.php]: Let's not add unit tests for platform or driver-specific behavior. Outside of the database, this SQL means nothing, so the database should be involved in producing coverage.

File: src/Platforms/SQLite/SQLiteMetadataProvider.php (hunk 3 above)
Comment (lazerg) [reply src/Platforms/SQLite/SQLiteMetadataProvider.php]: Fixed in 2cf231f: the conditional is gone, the parsed value is now assigned unconditionally. The only consumer checks `$options['without_rowid'] === true`, so an explicit `false` is safe.

File: tests/Functional/Schema/SQLiteSchemaManagerTest.php (hunk 4 above)
Comment (lazerg) [reply tests/Functional/Schema/SQLiteSchemaManagerTest.php]: Done in 2cf231f: the test is now parametrized with `#[TestWith]` over `true` and `false`.

File: tests/Platforms/SQLite/SQLiteMetadataProviderTest.php (hunk 5 above)
Comment (lazerg) [reply tests/Platforms/SQLite/SQLiteMetadataProviderTest.php]: Removed in 2cf231f. I moved the two parsing edge cases that were worth keeping into SQLiteSchemaManagerTest as functional tests running real DDL: `WITHOUT ROWID` followed by another table option, and a column named like the option.

**Comment by lazerg** [comment]
@derrabus whenever you get a chance, this is scoped down to WITHOUT ROWID only now and ready for another look.

**Review by derrabus — APPROVED** [review]

**Comment by derrabus** [comment]
Thank you!
